libraries/tock-cells/src/map_cell.rs

```diff
@@ -80,24 +162,69 @@ impl<T> MapCell<T> {
 
     /// Puts a value into the `MapCell`.
     pub fn put(&self, val: T) {
-        self.occupied.set(true);
-        unsafe {
-            ptr::write(self.val.get(), MaybeUninit::<T>::new(val));
-        }
+        // This will ensure the value as dropped
+        self.replace(val);
     }
 
     /// Replaces the contents of the `MapCell` with `val`. If the cell was not
     /// empty, the previous value is returned, otherwise `None` is returned.
-    pub fn replace(&self, val: T) -> Option<T> {
-        if self.is_none() {
-            self.put(val);
-            None
-        } else {
-            unsafe {
+    /// In the event the cell is currently borrowed, returns Err(AlreadyBorrowed)
+    pub fn try_replace(&self, val: T) -> Result<Option<T>, MapCellErr> {
+        match self.occupied.get() {
+            Uninit => {
+                unsafe {
+                    // Safety: Because we are in the Uninit state, we are not writing over anything
+                    // that needs to be dropped
+                    ptr::write(self.val.get(), MaybeUninit::<T>::new(val));
+                    self.occupied.set(Init)
+                }
+                Ok(None)
+            }
+            Init => unsafe {
+                // Safety: Because we are not in the InitBorrowed state, nothing is currently also
+                // referencing this value
                 let result: MaybeUninit<T> = ptr::replace(self.val.get(), MaybeUninit::new(val));
                 // `result` is _initialized_ and now `self.val` is now a new uninitialized value
-                Some(result.assume_init())
+                Ok(Some(result.assume_init()))
+            },
+            InitBorrowed => Err(AlreadyBorrowed),
+        }
+    }
+
+    /// Same as try_replace but panics if the cell is already borrowed
+    pub fn replace(&self, val: T) -> Option<T> {
+        match self.try_replace(val) {
+            Err(_) => replace_panic(),
+            Ok(v) => v,
+        }
+    }
+
+    /// Try borrow a mutable reference to the data contained in this cell
+    /// The type wrapped in the Option is a smart pointer to a T
+    /// Using this method rather than the callback based logic below will result in considerably
+    /// less code noise due to nested callbacks.
+    ///  # Examples
+    ///
+    /// ```
+    /// extern crate tock_cells;
+    /// use tock_cells::map_cell::MapCell;
+    /// let cell = MapCell::new(1234);
+    /// let x = &cell;
+    /// let y = &cell;
+    ///
+    /// *x.try_borrow_mut().unwrap() += 1;
+    ///
+    /// assert_eq!(y.take(), Some(1235));
+    /// ```
+    #[inline]
+    pub fn try_borrow_mut(&self) -> Result<MapCellRef<T>, MapCellErr> {
```

Comment:
...do we need this? I'm glancing at the CL and don't see a previous equivalent `borrow_mut`. It should simplify the necessary code in `try_replace`

Reply:
I agree, I do not think that this needs to be a `pub` interface, and we do not call this externally. If a more efficient function could be made used by not making this `pub` that works for me.